File: src/prepStats.R
prepStats = function(apps=NULL,vacs=NULL,assigned=NULL,prio=NULL,iteration=NULL){
  cat(paste0('\n        Prep stats'))
  
  ## Get considered:
  d3 = merge(apps,assigned[,.(applicant_id,ass_priority_profile=trunc(score))],by='applicant_id',all.x=T)
  d3[is.na(ass_priority_profile),ass_priority_profile:=0]
  d3 = d3[ass_priority_profile >= priority_profile,] ## If more preferred or was assigned
  d3 = d3[,.(n_considered = .N),by=.(program_id,quota_id,priority_profile)]
  d3 = d3[order(program_id,quota_id,priority_profile)]
  
  tot_apps = apps[,.(total_applicants=.N),by=.(program_id,quota_id)]
  
  pp = merge(prio,assigned[,.(assigned=.N),by=.(program_id,quota_id,priority_profile=trunc(score))],all.x=T)
  pp = merge(pp,tot_apps,by=c('program_id','quota_id'),all.x=T)
  pp = merge(pp,d3,by=c('program_id','quota_id','priority_profile'),all.x=T)
  pp[is.na(pp)] = 0
  pp = pp[order(program_id,quota_id,-priority_profile)]
  pp[,cum_ass:=cumsum(assigned),by=.(program_id,quota_id)]
  pp = merge(pp,vacs[,.(seats=max(seat_order)),by=.(program_id,quota_id)],by=c('program_id','quota_id'),all.x=T)
  pp[is.na(pp)] = 0
  pp[,iter:=iteration]
  
  ## Create variables:
  pp = pp[order(iter,program_id,quota_id,-priority_profile)]
  pp[,assigned := shift(cum_ass,n=1,type='lag'),by=.(iter,program_id,quota_id)]
  pp[is.na(pp)] = 0
  pp[,assigned := cum_ass - assigned]
  pp[,cum_ass := cum_ass - assigned]
  pp[,total_seats := seats]
  pp[,seats:=seats - cum_ass]
  
  return(pp)
}

File: src/prepStats.py
def prepStats(apps, vacs, assigned, prio, iteration):
    d3 = apps.merge(assigned[['applicant_id', 'ass_priority_profile']], on='applicant_id', how='left')
    d3['ass_priority_profile'].fillna(0, inplace=True)
    d3 = d3[d3['ass_priority_profile'] >= d3['priority_profile']]
    d3 = d3.groupby(['program_id', 'quota_id', 'priority_profile']).size().reset_index(name='n_considered')
    
    tot_apps = apps.groupby(['program_id', 'quota_id']).size().reset_index(name='total_applicants')
    
    pp = prio.merge(assigned.groupby(['program_id', 'quota_id', 'priority_profile']).size().reset_index(name='assigned'), on=['program_id', 'quota_id', 'priority_profile'], how='left')
    pp = pp.merge(tot_apps, on=['program_id', 'quota_id'], how='left')
    pp = pp.merge(d3, on=['program_id', 'quota_id', 'priority_profile'], how='left')
    pp.fillna(0, inplace=True)
    pp['cum_ass'] = pp.groupby(['program_id', 'quota_id'])['assigned'].cumsum()
    pp = pp.merge(vacs.groupby(['program_id', 'quota_id']).agg(seats=('seat_order', 'max')), on=['program_id', 'quota_id'], how='left')
    pp.fillna(0, inplace=True)
    pp['iter'] = iteration
    pp.sort_values(by=['iter', 'program_id', 'quota_id', -pp['priority_profile'].astype(int)], inplace=True)
    pp['assigned'] = pp.groupby(['iter', 'program_id', 'quota_id'])['cum_ass'].shift().fillna(0)
    pp['assigned'] = pp['cum_ass'] - pp['assigned']
    pp['cum_ass'] -= pp['assigned']
    pp['total_seats'] = pp['seats']
    pp['seats'] -= pp['cum_ass']
    return pp
